### PR TITLE
Supervised Learning Algorithms

### PR DESCRIPTION
**Pull request by lalith1403**
short_description: Introduction to supervised learning algorithms such as classification and regression algorithms.

### CODE REVIEW

File: articles/supervised-learning-algorithms/index.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    # Split into training and test set
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=42, stratify=y)
+
+    linear_regression = LinearRegression()
+    linear_regression.fit(X_train,y_train)
+
+    print(linear_regression.score(X_test,y_test))
+    ```
+
+The output of the print statement will be 0.55. The scoring metric used for linear regression is the R^2 metric. Pronounced as R-squared, it tells us about the effectiveness of the curve fitting on the graph. The closer the value of R-squared is to 1, the higher chances of good curve fitting. Curve fitting may lead to overfitting when the number of features considered is less. Overfitting refers to a scenario where the model performs very well on the data it has seen. But its performance drops when it works on unseen data. Underfitting is also a possibility when we don't have sufficient data to train the model(s) with.
```
Comment (hectorkambow) [inline articles/supervised-learning-algorithms/index.md:151]: This paragraph confused me a bit - when it came to the curve fitting - it may have been me - since Im not familiar with models - but the mention of a curve - made me pause for a moment and ask what curve are they talking about - it took me a second - to understand all these models are represented on graphs. 
The overfitting is kind of confusion - is there any way to include a bit more about that. A few sentences explaining the curve fitting -

File: articles/supervised-learning-algorithms/index.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=42, stratify=y)
+
+    # Create a SVM classifier
+    classifier = svm.SVC(kernel='poly')
+    classifier.fit(X_train, y_train)
+    print(classifier.score(X_test,y_test))
+   ```
+
+### Regression
+Regression algorithms are another subset of machine learning algorithms used to predict continuous numeric responses. As seen in an earlier example, predicting house rent given different factors is an example of regression. Let's look at the regression algorithm and use ridge regression as an example.
```
Comment (hectorkambow) [inline articles/supervised-learning-algorithms/index.md:125]: ? @lalith1403  - while reading this paragraph - I see that you mentioned ridge regression ? But then below you talk about linear regression? 
not sure if its the same thing - but that was a bit confusing

File: articles/supervised-learning-algorithms/index.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+1. K-Nearest Neighbors(KNN): KNN is an algorithm that works on creating a decision boundary based on distance metrics. Distance metrics define and parameterize distance. There are various distance metrics such as Euclidean distance, Manhattan distance, etc.
+
+Manhattan distance takes the absolute distance between two points, whereas Euclidean distance takes the square root of the sum of the squares of the distance between the two points.
+
+![distance metrics](/engineering-education/supervised-learning-algorithms/distance.jpg)
+
+*[Image Source](https://www.ieee.ma/uaesb/pdf/distances-in-classification.pdf)*
+
+All machine learning algorithms have hyperparameters to deal with. In K-NN, the parameter is `k`. It is initialized to an integer depending on the number of classes in the dataset known apriori to fitting. `k` signifies the number of nearest points the algorithm considers while creating decision boundaries.
```
Comment (hectorkambow) [inline articles/supervised-learning-algorithms/index.md:50]: "dataset known apriori to fitting." 
This bit was a bit confusing  - just not sure what apriori means - is it a ML term? If so it may be useful to and a link or explain a bit. 🚀

File: articles/supervised-learning-algorithms/index.md (hunk 2 above)
Comment (lalith1403) [reply articles/supervised-learning-algorithms/index.md:125]: Sorry about that... I meant linear regression.

File: articles/supervised-learning-algorithms/index.md (hunk 3 above)
Comment (lalith1403) [reply articles/supervised-learning-algorithms/index.md:50]: apriori means prior to.... I will reword it

Comment (hectorkambow) [reply articles/supervised-learning-algorithms/index.md]: thank you - I thought that what it may mean - but when I looked it up a ML term also popped up so I was sure. 👍

**Comment by hectorkambow** [comment]
👋 Marcus @mhoulmiere (happy pre-Friday) wanted to ask if you could give this "Supervised Learning Algorithms" article a tech review when you have a moment - not sure if it's in your wheelhouse.

Let me know if I should re-assign or have any questions. 

Thank you again 🚀

**Comment by hectorkambow** [comment]
@kurt-sectionio Thanks again, Kurt - 🚀 
I learned a bit about supervised and unsupervised learning  🤖 
Let me know if you have any ?s

**Review by kurt-sectionio — CHANGES_REQUESTED** [review]
Very good. See notes. The main thing that needs attention is the distance measures. Needs a little work (notes below).

line 22: {soapbox moment) You don't have to change it, but I do not like this definition of machine learning. Why not? You give examples below of machine learning that do not 'automatically improve through experience'. If that is a requirement of machine learning, then you are not discussing any machine learning in your article. I personally define ML as using a particular collection of algorithms to extract patterns from data. That is exactly what your examples are doing. I would define statistics similarly, however that field uses a different collection of algorithms. I also don't consider ML as a subset of AI. Rather, I see AI as a set of activities that uses ML and some other technologies (usually interactive ML systems) to achieve automated improvement. So I think the definition you provide is more suitable for AI if we are trying to distinguish AI from ML. I don't seek AGI in my AI projects, I seek the benefits of automating knowledge/judgement tasks, synthesizing and operationalizing the pattern-extraction coming from the base ML algorithms. Again--that's just my personal view.

line 38: 'A','B','C' might be better than 1, 2, 3  due to possible confusion between categorical classifications and numerical values.

line 44: Check you distance definitions, I don't think they convey the essence of either distance measure. I find the "absolute distance" reference to Manhattan distance to be ambiguous and potentially misleading. Manhattan distance is a series of straight lines connected with right angles between 2 points. It is so-named because it is exemplified by the distance one walks/drives along city streets to get from point A to point A. Euclidean distance is a simple, straight line, the shortest such line between two points. When using typical Cartesian coordinates (e.g., x and y), you compute the shortest straight-line distance between them using the Pythagorean Formula, which you describe.

It may be enough to mention "distance" measures in general. KNN depends on a notion of "nearness" (proximity). There are many ways to model proximity. One need not know, e.g., manhattan distance to understand your real point here.

**Review by kurt-sectionio — COMMENTED** [review]
You changed the text of the quoted definition of ML? How can that be? I went to the link and the term 'pattern' does not show up there. This is a bad thing. If you wnat to use my definition, just use it without quotes and drop the reference to Tom

**Comment by lalith1403** [comment]
> Very good. See notes. The main thing that needs attention is the distance measures. Needs a little work (notes below).
> 
> line 22: {soapbox moment) You don't have to change it, but I do not like this definition of machine learning. Why not? You give examples below of machine learning that do not 'automatically improve through experience'. If that is a requirement of machine learning, then you are not discussing any machine learning in your article. I personally define ML as using a particular collection of algorithms to extract patterns from data. That is exactly what your examples are doing. I would define statistics similarly, however that field uses a different collection of algorithms. I also don't consider ML as a subset of AI. Rather, I see AI as a set of activities that uses ML and some other technologies (usually interactive ML systems) to achieve automated improvement. So I think the definition you provide is more suitable for AI if we are trying to distinguish AI from ML. I don't seek AGI in my AI projects, I seek the benefits of automating knowledge/judgement tasks, synthesizing and operationalizing the pattern-extraction coming from the base ML algorithms. Again--that's just my personal view.
> 
> line 38: 'A','B','C' might be better than 1, 2, 3 due to possible confusion between categorical classifications and numerical values.
> 
> line 44: Check you distance definitions, I don't think they convey the essence of either distance measure. I find the "absolute distance" reference to Manhattan distance to be ambiguous and potentially misleading. Manhattan distance is a series of straight lines connected with right angles between 2 points. It is so-named because it is exemplified by the distance one walks/drives along city streets to get from point A to point A. Euclidean distance is a simple, straight line, the shortest such line between two points. When using typical Cartesian coordinates (e.g., x and y), you compute the shortest straight-line distance between them using the Pythagorean Formula, which you describe.
> 
> It may be enough to mention "distance" measures in general. KNN depends on a notion of "nearness" (proximity). There are many ways to model proximity. One need not know, e.g., manhattan distance to understand your real point here.

Thank you @kurt-sectionio ... the review was detailed and highly informative. I have made the necessary changes. :)

**Comment by kurt-sectionio** [comment]
Please fix the ML definition per my comment above.

**Review by kurt-sectionio — APPROVED** [review]
Thanks. Looks good.
